File: 03_Linear_equations/Scripts_it/04_iterative-methods-mfiles/solveJacobi.m
function [x,it_jac] = solveJacobi(A,b,tol)
% Solves a linear system using the Jacobi method

% Set default error
if nargin < 3
    tol = 1e-1;
end

x = 25*ones(size(b));   % Set initial guess to b
N = length(A);          % Number of equations
it_jac = 1;             % Init number of iterations

% While not converged or max_it not reached
while ( norm(A*x-b, 2) > tol && it_jac < 1000 )
    x_old = x;
    for i=1:N
        s = 0;
        for j = 1:N
            if (j ~= i)
                % Sum off-diagonal*x_old
                s = s+A(i,j)*x_old(j);
            end
        end
        % Compute new x value
        x(i) = (b(i)-s)/A(i,i);
    end
    % Increate number of iterations
    it_jac = it_jac+1;
end


% Uncomment this part to use array-indices. Remember to comment the part
% above!
% 
% while ( norm(A*x-b, 2) > tol && it_jac < 1000 )
%     x_old = x;
%     for i = 1:N
%         x(i) = (1/A(i,i))*((b(i) - A(i,[1:i-1,i+1:N]) * x_old([1:i-1,i+1:N])));
%     end
%     it_jac = it_jac + 1;
% end

it_jac

File: scripts/used_by_students/03c_iterative-methods-mfiles_v4/solveJacobi.py
import numpy as np
from numpy.linalg import norm

# Define the function
def solve_jacobi(A, b, tol=1e-2):
    
    # Set initial guess
    x = b + 1e-16
        
    # Initialize variables
    x_diff = 1
    N = len(A)
    it_jac = 1
    
    # While not converged or max_it not reached
    while (x_diff > tol and it_jac < 1000):
        x_old = x.copy()
        for i in range(N):
            s = 0
            for j in range(N):
                if (j != i):
                    # Sum off-diagonal*x_old
                    s += A[i,j] * x_old[j]
            # Compute new x value
            x[i] = (b[i] - s) / A[i,i]
            
        # Increase number of iterations
        it_jac += 1
        x_diff = norm((x - x_old), 2)
        
    # Print number of iterations
    print(it_jac)
    
    return x, it_jac


A = np.array([[4, -1, 0, 0], [-1, 4, -1, 0], [0, -1, 4, -1], [0, 0, -1, 3]])
b = np.array([15, 10, 10, 10])
x = np.linalg.solve(A,b)
x_ = solve_jacobi(A,b, 1e-6)[0]
print(x, x_)